src/extensions/rolemanagement/member_search.py

```diff
@@ -0,0 +1,131 @@
+#   Copyright 2020-present Michael Hall
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from __future__ import annotations
+
+import asyncio
+from typing import AsyncIterator, Callable, Optional, Set
+
+import discord
+
+
+async def _member_yielder(
+    members: Set[discord.Member],
+) -> AsyncIterator[discord.Member]:
+
+    # While the check function isn't expensive,
+    # we can ensure this doesn't create issues
+    # when applied to larger member lists
+    # with periodic yielding of the event loop.
+
+    for index, member in enumerate(members):
+        if not index % 1000:
+            await asyncio.sleep(0)
+        yield member
+
+
+async def search_filter(
+    members: Set[discord.Member], query: dict
+) -> Set[discord.Member]:
+    """
+    Reusable filter
+    """
+
+    # sleep(0) use is used to yield control of the event loop for a cycle
+
+    if query["everyone"]:
+        return members
+
+    if all_roles := query["all"]:
+        members.intersection_update(*(role.members for role in all_roles))
+        await asyncio.sleep(0)
+
+    if none_roles := query["none"]:
+        members.difference_update(*(role.members for role in none_roles))
+        await asyncio.sleep(0)
+
+    if any_roles := query["any"]:
+        any_set: Set[discord.Member] = set()
+        any_set.update(*(role.members for role in any_roles))
+        members.intersection_update(any_set)
+        await asyncio.sleep(0)
+
+    filter_obj = _non_set_filter(query)
+
+    return {m async for m in _member_yielder(members) if filter_obj(m)}
+
+
+def _non_set_filter(query: dict) -> Callable[[discord.Member], bool]:
```

Comment:
Take or leave this one: Change this to return `Optional[Callable[[discord.Member], bool]]`, None when the conditions are empty, then change where it's used to save iterations when it's only used to check membership in specific roles.